### PR TITLE
[OPENCOST-1777] Support n > 1 rate codes for AWS on demand

### LINKED ISSUE #1777: Incorrect OnDemandRateCodeCn configuration
https://github.com/opencost/opencost/blob/3507d95d08803ef541ce3f147b40db500d10c1b6/pkg/cloud/awsprovider.go#L304

The current correct OnDemandRateCodeCn is 5Y9WH78GDR

image

### PR DESCRIPTION
**Pull request by ameijer**
## What does this PR change?
* Implements support for multiple rate codes as per #1777 . Also adds unit testing and limited refactor of the price retrieval and parsing logic. 

## Does this PR relate to any other PRs?
* None. 

## How will this PR impact users?
* AWS US users should be unaffected. CN AWS users should now have all rate codes supported. 

## Does this PR address any GitHub or Zendesk issues?
* Closes #1777 

## How was this PR tested?
* Unit tests were added. Checked for regressions by running the app and confirming no errors.

## Does this PR require changes to documentation?
* no. 

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
*

### CODE REVIEW

**Comment by mattray** [comment]
Is this associated with https://github.com/opencost/opencost/issues/1777?

**Comment by AjayTripathy** [comment]
Yes!

File: pkg/cloud/awsprovider.go [hunk 1]
```diff
@@ -1427,7 +1410,7 @@ func (aws *AWS) getAWSAuth(forceReload bool, cp *CustomPricing) (string, string)
 	}
 
 	// 3. Fall back to env vars
-	if env.GetAWSAccessKeyID() == "" || env.GetAWSAccessKeyID() == "" {
+	if env.GetAWSAccessKeyID() == "" || env.GetAWSAccessKeySecret() == "" {
```
Comment (ameijer) [inline pkg/cloud/awsprovider.go:1416]: @AjayTripathy my IDE flagged this as a no-op, assiming the intent here was to check both access key ID and secret key?

File: pkg/cloud/awsprovider.go [hunk 2]
```diff
@@ -359,34 +364,6 @@ var locationToRegion = map[string]string{
 	"AWS GovCloud (US-West)":    "us-gov-west-1",
 }
 
-var regionToBillingRegionCode = map[string]string{
```
Comment (ameijer) [inline pkg/cloud/awsprovider.go:362]: this appears to be unused

File: pkg/cloud/awsprovider.go [hunk 3]
```diff
@@ -923,19 +900,75 @@ func (aws *AWS) DownloadPricingData() error {
...
 
-	aws.Pricing = make(map[string]*AWSProductTerms)
 	aws.ValidPricingKeys = make(map[string]bool)
-	skusToKeys := make(map[string]string)
 
 	resp, pricingURL, err := aws.getRegionPricing(nodeList)
 	if err != nil {
 		return err
 	}
+	aws.populatePricing(resp, inputkeys)
```
Comment (ameijer) [inline pkg/cloud/awsprovider.go]: the in-line logic where we read the body and parsed it into an AWS object was moved to its own function to both improve testability and simplify this function

File: pkg/cloud/awsprovider.go [hunk 4]
```diff
@@ -923,19 +900,53 @@ func (aws *AWS) DownloadPricingData() error {
...
 
-	aws.Pricing = make(map[string]*AWSProductTerms)
 	aws.ValidPricingKeys = make(map[string]bool)
-	skusToKeys := make(map[string]string)
 
 	resp, pricingURL, err := aws.getRegionPricing(nodeList)
 	if err != nil {
 		return err
 	}
+	err = aws.populatePricing(resp, inputkeys)
```
Comment (ameijer) [inline pkg/cloud/awsprovider.go:909]: the parsing was moved to its own function to facilitate testing and to simplify this 'runner' function

File: pkg/cloud/awsprovider.go (hunk 1 above)
Comment (AjayTripathy) [reply pkg/cloud/awsprovider.go:1416]: Yes, good find thank you.

**Review by AjayTripathy — APPROVED** [review]

**Pull request merged**